### PR TITLE
chore: Email COA's from Sales Invoice

### PR DESCRIPTION
**Pull request by harshmule-epi**
> [Asana task](https://app.asana.com/0/1200414497004131/1200166130781170/f)

> To Do: to mass email COA's from the sales invoice doctype

### CODE REVIEW

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 1]
```diff
@@ -1591,3 +1591,31 @@ def create_invoice_discounting(source_name, target_doc=None):
 def get_item_details(item_code):
 	item = frappe.get_doc("Item", item_code).as_dict()
 	return item
+
+@frappe.whitelist()
+def email_coas(docname):
+	sales_invoice = frappe.get_doc("Sales Invoice", docname).as_dict()
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
	sales_invoice = frappe.get_doc("Sales Invoice", docname)
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 2]
```diff
@@ -1591,3 +1591,31 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+		if item.get("certificate_of_analysis")]
+
+	if not coas:
+		frappe.msgprint(_("No Certificates of Analysis attached"))
+		return
+
+	attachments = []
+	for item in sales_invoice.get("items"):
+		coa_file_id = frappe.db.get_value("File", {"file_url": item.certificate_of_analysis}, "name")
+		attachments.append({"fid": coa_file_id})
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: commonify the loops, no need for multiple for loops

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py (hunk 2 above)
Comment (harshmule-epi) [reply erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: > commonify the loops, no need for multiple for loops

done the changes!

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 3]
```diff
@@ -1591,3 +1591,26 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+	sales_invoice = frappe.get_doc("Sales Invoice", docname)
+	attachments = []
+	if not sales_invoice.get("contact_email"):
+		frappe.throw(_("No contact email found"))
+	for item in sales_invoice.get('items'):
+		item.certificate_of_analysis = frappe.db.get_value("Batch", item.batch_no, "certificate_of_analysis")
+		coas = item.get("certificate_of_analysis")
+		if not coas:
+			frappe.msgprint(_("No Certificates of Analysis attached"))
+			return
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1606]:
```suggestion
			continue
```
do you want to break the loop if even one of the items do not have a COA?

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py (hunk 3 above)
Comment (harshmule-epi) [reply erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1606]: as per the DOD the same functionality is done in Delivery Note, so i tried to follow the same procedure, as coas is not coming so they are breaking or returning the function null.

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**